pkg/nfd-master/nfd-master.go

```diff
@@ -440,10 +451,114 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
 			klog.Errorf("failed to advertise labels: %v", err)
 			return &pb.SetLabelsReply{}, err
 		}
+
+		// set taints
+		var taints []corev1.Taint
+		if m.args.EnableTaints {
+			taints = crTaints
+		}
+
+		// Call setTaints even though the feature flag is disabled. This
+		// ensures that we delete NFD owned stale taints when flag got
+		// turned off.
+		err = m.setTaints(c, cli, taints, r.NodeName)
+		if err != nil {
+			return &pb.SetLabelsReply{}, err
+		}
 	}
 	return &pb.SetLabelsReply{}, nil
 }
 
+// setTaints sets note taints and annotations based on the taints passed via
+// nodeFeatureRule custom resorce. If empty list of taints is passed, currently
+// NFD owned taints and annotations are removed from the node.
+func (m *nfdMaster) setTaints(c context.Context, cli *kubernetes.Clientset, taints []corev1.Taint, nodeName string) error {
+	// Fetch the node object.
+	node, err := m.apihelper.GetNode(cli, nodeName)
+	if err != nil {
+		return err
+	}
+
+	// De-serialize the taints annotation into Taint type for comparision below.
+	annotations := node.GetAnnotations()
+	oldTaints := []corev1.Taint{}
+	var val string
+	if val, ok := annotations[nfdv1alpha1.NodeTaintsAnnotation]; ok {
+		sts := strings.Split(val, ",")
+		corev1Taints, _, err := taintutils.ParseTaints(sts)
+		if err != nil {
+			return err
+		}
+		for _, ct := range corev1Taints {
+			oldTaints = append(oldTaints,
+				corev1.Taint{
+					Key:    ct.Key,
+					Value:  ct.Value,
+					Effect: ct.Effect})
+		}
+	}
+
+	// Find out the diff between the old (from taints annotation) and
+	// new taints (from NodeFeatureRule) from which we get the taints
+	// to remove from the node.
+	taintsToRemove := []corev1.Taint{}
+	for _, taint := range oldTaints {
+		if !taintutils.TaintExists(taints, &taint) {
+			taintsToRemove = append(taintsToRemove, taint)
+		}
+	}
+
+	// We have taintsToRemove that we found from the annotation and here
+	// we first check if the taints still exist on the node and if yes
+	// remove it, otherwise move on.
+	for _, taintToRemove := range taintsToRemove {
+		if taintutils.TaintExists(node.Spec.Taints, &taintToRemove) {
+			newTaints, removed := taintutils.DeleteTaint(node.Spec.Taints, &taintToRemove)
+			if !removed {
+				return fmt.Errorf("failed to delete %q taint", taintToRemove.ToString())
+			}
+			node.Spec.Taints = newTaints
+		}
+	}
+
+	// Add new taints.
+	var nodeTaints []corev1.Taint
+	if len(taints) > 0 {
+		for _, taint := range taints {
+			if !taintutils.TaintExists(node.Spec.Taints, &taint) {
+				nodeTaints = append(nodeTaints, taint)
+			}
+			continue
+		}
+		node.Spec.Taints = append(node.Spec.Taints, nodeTaints...)
+	}
+
+	// Serialize the new taints into string and update the annotation
+	// with that string.
+	taintStrs := []string{}
+	oldAnnotation := annotations[nfdv1alpha1.NodeTaintsAnnotation]
+	if len(taints) > 0 {
+		for _, taint := range taints {
```

Comment:
nit: declare variables where they are used
```suggestion
	oldAnnotation := annotations[nfdv1alpha1.NodeTaintsAnnotation]
	if len(taints) > 0 {
		taintStrs := []string{}
		for _, taint := range taints {
```